tensorboard/manager_e2e_test.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""End-to-end tests for `tensorboard.manager`."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import contextlib
+import datetime
+import errno
+import json
+import os
+import pipes
+import signal
+import subprocess
+import tempfile
+import textwrap
+
+import six
+from six.moves import urllib
+import tensorflow as tf
+
+try:
+  # python version >= 3.3
+  from unittest import mock  # pylint: disable=g-import-not-at-top
+except ImportError:
+  import mock  # pylint: disable=g-import-not-at-top,unused-import
+
+from tensorboard import manager
+
+
+class ManagerEndToEndTest(tf.test.TestCase):
+
+  def setUp(self):
+    super(ManagerEndToEndTest, self).setUp()
+
+    # Spy on subprocesses spawned so that we can kill them.
+    self.popens = []
+    class PopenSpy(subprocess.Popen):
```

Comment:
I might be wrong but isn't this a mock, not a spy?

Reply:
Maybe? I don’t know. “Spy” seems more appropriate to me because I _am_
trying to spy on the behavior but I’m _not_ actually mocking anything
out (we delegate to the real implementation, and actually do spawn
subprocesses). If it’d make you happy for this to say `PopenMock`
instead, I can change it.

Frankly, I have never carefully looked into the Official Terminology™ of
“mocks, stubs, and spies” or “libraries and frameworks” or similar
distinctions because I don’t see how they’re important or useful. :-)